### PR TITLE
wolfTPM pubkey storage with policy based access restriction

### PR DESCRIPTION
**Pull request by jpbland1**
this update uses the tpm to retreive the public key used to validate the image that will boot and restricts access to that key by tpm policy. when the image is updated it's signature is used to extend the PCR and when the image is loaded it's signature must match what was sealed in order to get the public key from the tpm. enabling this option is done by setting WOLFBOOT_TPM_KEYSTORE in .config

### CODE REVIEW

File: src/image.c [hunk 1]
```diff
@@ -36,6 +35,17 @@
 #include "wolftpm/tpm2.h"
 #include "wolftpm/tpm2_wrap.h"
 static WOLFTPM2_DEV wolftpm_dev;
+
+#ifdef WOLFBOOT_TPM_KEYSTORE
+static WOLFTPM2_SESSION wolftpm_session;
+static uint32_t wolftpmPcrArray[1] = {16};
```
Comment (dgarske) [inline src/image.c:41]: Should this be an array of uint8_t?

File: src/image.c [hunk 2]
```diff
@@ -36,6 +35,17 @@
...
+static const uint32_t wolftpmKeystoreIndex = 0x01800200;
```
Comment (dgarske) [inline src/image.c:42]: Can you make all of these overridable build-time macros?

File: src/image.c [hunk 3]
```diff
@@ -770,8 +786,181 @@ int wolfBoot_tpm2_init(void)
     if (rc != 0)  {
         return rc;
     }
+
+#ifdef WOLFBOOT_TPM_KEYSTORE
+    /* start a policy session with parameter encryption */
+    rc = wolfTPM2_StartSession(&wolftpm_dev, &wolftpm_session, NULL, NULL,
+        TPM_SE_POLICY, TPM_ALG_CFB);
+    if (rc != 0)
```
Comment (dgarske) [inline src/image.c:794]: Maybe use `TPM_RC_SUCCESS`?

File: src/image.c [hunk 4]
```diff
@@ -770,8 +786,181 @@ int wolfBoot_tpm2_init(void)
...
+        return rc;
+
+    /* set the auth session for the device */
+    rc = wolfTPM2_SetAuthSession(&wolftpm_dev, 0, &wolftpm_session,
+        (TPMA_SESSION_decrypt | TPMA_SESSION_encrypt | TPMA_SESSION_continueSession));
+    if (rc != 0)
```
Comment (dgarske) [inline src/image.c:800]: Maybe use `TPM_RC_SUCCESS`?

File: src/image.c [hunk 5]
```diff
@@ -770,8 +786,181 @@ int wolfBoot_tpm2_init(void)
...
+    uint8_t* pubkeyHint;
+    uint8_t* pubkey;
+    uint8_t* policySignature;
+    uint32_t tpmPubkeySz = KEYSTORE_PUBKEY_SIZE_ECC256;
+    int keySlot;
+    int ret;
+    uint16_t policySignatureSz;
+    uint16_t imageSignatureSz;
+    uint16_t pubkeyHintSize;
+    uint8_t tpmPubkey[KEYSTORE_PUBKEY_SIZE_ECC256];
```
Comment (dgarske) [inline src/image.c:822]: Is forcing ECC256 okay? `KEYSTORE_PUBKEY_SIZE_ECC256`?

File: src/image.c [hunk 6]
```diff
@@ -770,8 +786,181 @@ int wolfBoot_tpm2_init(void)
...
+
+    XMEMSET(&tpmKey, 0, sizeof(tpmKey));
+
+    /* find the keyslot of the public key */
+    pubkeyHintSize = get_header(newImg, HDR_PUBKEY, &pubkeyHint);
+    if (pubkeyHintSize != WOLFBOOT_SHA_DIGEST_SIZE)
+        return -1; /* Invalid hash size for public key hint */
```
Comment (dgarske) [inline src/image.c:829]: Does wolfBoot not have better return codes?

File: src/image.c [hunk 7]
```diff
@@ -770,8 +786,181 @@ int wolfBoot_tpm2_init(void)
...
+
+    /* get the backupImg signature */
+    imageSignatureSz = get_header(backupImg, HDR_SIGNATURE, &imageSignature);
+    if (imageSignatureSz != IMAGE_SIGNATURE_SIZE)
+       return -1;
+
+    /* clear out the policy digest */
+    ret = wolfTPM2_PolicyRestart(wolftpm_session.handle.hndl);
+    if (ret != TPM_RC_SUCCESS)
+        return -1;
```
Comment (dgarske) [inline src/image.c:848]: Maybe just convert the TPM codes to negative?

File: src/image.c [hunk 8]
```diff
@@ -770,8 +786,181 @@ int wolfBoot_tpm2_init(void)
...
+    rc = wolfTPM2_SetAuthSession(&wolftpm_dev, 0, &wolftpm_session,
+        (TPMA_SESSION_decrypt | TPMA_SESSION_encrypt | TPMA_SESSION_continueSession));
+    if (rc != 0)
+        return rc;
+#endif
+    return 0;
+}
+
+#ifdef WOLFBOOT_TPM_KEYSTORE
+/* when this is called, update is the backup image and boot is the new */
```
Comment (dgarske) [inline src/image.c:807]: If this is only supported with ECC256 please document.

File: src/image.c [hunk 9]
```diff
@@ -770,8 +786,181 @@ int wolfBoot_tpm2_init(void)
...
+
+    /* unseal the NV pubkey */
+    ret = wolfTPM2_UnsealWithAuthSigNV(&wolftpm_dev, tpmKey, &wolftpm_session,
+        TPM_ALG_SHA256, (word32*)wolftpmPcrArray, sizeof(wolftpmPcrArray), NULL,
+        0, policySignature, policySignatureSz, wolftpmKeystoreIndex,
+        wolftpmPolicyDigestIndex, tpmPubkey, (word32*)&tpmPubkeySz);
+    if (ret != TPM_RC_SUCCESS)
+        return -1;
+
+    /* unload the intermidate key */
```
Comment (dgarske) [inline src/image.c:951]: `intermediate`

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/image.c (hunk 1 above)
Comment (dgarske) [reply src/image.c:41]: Array of uint8_t?

File: src/image.c (hunk 7 above)
Comment (dgarske) [reply src/image.c:849]: Convert TPM codes to negative?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: options.mk [hunk 10]
```diff
@@ -5,6 +5,16 @@ ifeq ($(MEASURED_BOOT),1)
   CFLAGS+=-D"WOLFBOOT_MEASURED_PCR_A=$(MEASURED_PCR_A)"
 endif
 
+## use tpm to store true pubkey
+ifeq ($(WOLFBOOT_TPM_KEYSTORE),1)
```
Comment (dgarske) [inline options.mk:9]: Can you create an example .config that has `WOLFBOOT_TPM_KEYSTORE`? For testing?

File: src/image.c [hunk 11]
```diff
@@ -36,6 +35,15 @@
 #include "wolftpm/tpm2.h"
 #include "wolftpm/tpm2_wrap.h"
 static WOLFTPM2_DEV wolftpm_dev;
+
+#ifdef WOLFBOOT_TPM_KEYSTORE
+static WOLFTPM2_SESSION wolftpm_session;
+static uint8_t wolftpmPcrArray[1] = {16};
```
Comment (dgarske) [inline src/image.c:41]: This PR index should be a build-time overridable value.

File: tools/keytools/sign.c [hunk 12]
```diff
@@ -711,6 +715,14 @@ static int make_header_ex(int is_diff, uint8_t *pubkey, uint32_t pubkey_sz,
             goto failure;
         }
         memset(signature, 0, CMD.signature_sz);
+        if (CMD.manual_sign && CMD.policy_sign) {
```
Comment (dgarske) [inline tools/keytools/sign.c:718]: Can you please document these new signing tools arguments? Not clear how to use these?

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]
Looks good. @billphipps is going to test on customer HW.

File: options.mk (hunk 10 above)
Comment (danielinux) [reply options.mk:9]: +1 on this, ideally if it runs on STM32F407 so I can integrate with automated TPM tests on Jenkins

**Comment by danielinux** [comment]
@jpbland1 please add some documentation in docs/Target.md. Create a new section under the STM32F407 target (based on a new config `config/examples/stm32f4-tpm-keystore.config`  including the options needed). Please provide all the information about provisioning the key in the TPM to test the feature

**Review by danielinux — DISMISSED** [review]

**Review by dgarske — APPROVED** [review]
PR looks good. I'd like to git this into repo soon to start leveraging. Followup with new PR's for issues.

**Comment by dgarske** [comment]
Merging this now. Only boot with unseal tested, not an update scenario. Also needs instructions on how to preseal, since it requires doing manual signing and the extra key. That will come in future PR.

**Pull request merged**